File: C++/Rotate-List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    int getlength(ListNode* head) {
        int length = 0;
        while (head != NULL) {
            length ++;
            head = head->next;
        }
        return length;        
    }
    ListNode* rotateRight(ListNode* head, int k) {
        int len = getlength(head);
        if (len == 0) return head;
        if (k % len == 0) return head;
        k = k % len;
        ListNode *dummy = new ListNode(-1);
        dummy->next = head;
        ListNode *cur = dummy;
        for (int i = 0; i < k; i ++) {
            cur = cur->next;
        }
        ListNode *slow = dummy;
        while (cur->next != NULL) {
            cur = cur->next;
            slow = slow->next;
        }
        cur->next = dummy->next;
        dummy->next = slow->next;
        slow->next = NULL;
        return dummy->next;
    }
};

File: Python/Rotate-List.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    # @param head, a ListNode
    # @param k, an integer
    # @return a ListNode
    def rotateRight(self, head, k):
        if head == None or k == 0:
            return head
        node = head
        length = 0
        while node.next != None:
            node = node.next
            length += 1
        length += 1
        if length == 1 or k % length == 0:
            return head
        cnt = length - (k % length) - 1
        node1 = head
        for i in range(cnt):
            node1 = node1.next
        node2 = node1.next
        node1.next = None
        node.next = head
        return node2